---
title: "Pertussis resurgence mini-project"
author: "jack olmstead"
format: html
---

# Web scraping

I will extract the Pertussis epi data from the CDC's table using the datapasta package.

-   https://www.cdc.gov/pertussis/surv-reporting/cases-by-year.html

```{r echo=FALSE}
cdc <- data.frame(
  year = c(1922L,
           1923L,1924L,1925L,1926L,1927L,1928L,
           1929L,1930L,1931L,1932L,1933L,1934L,1935L,
           1936L,1937L,1938L,1939L,1940L,1941L,
           1942L,1943L,1944L,1945L,1946L,1947L,1948L,
           1949L,1950L,1951L,1952L,1953L,1954L,
           1955L,1956L,1957L,1958L,1959L,1960L,
           1961L,1962L,1963L,1964L,1965L,1966L,1967L,
           1968L,1969L,1970L,1971L,1972L,1973L,
           1974L,1975L,1976L,1977L,1978L,1979L,1980L,
           1981L,1982L,1983L,1984L,1985L,1986L,
           1987L,1988L,1989L,1990L,1991L,1992L,1993L,
           1994L,1995L,1996L,1997L,1998L,1999L,
           2000L,2001L,2002L,2003L,2004L,2005L,
           2006L,2007L,2008L,2009L,2010L,2011L,2012L,
           2013L,2014L,2015L,2016L,2017L,2018L,
           2019L
           ),
  cases = c(107473,
           164191,165418,152003,202210,181411,
           161799,197371,166914,172559,215343,179135,
           265269,180518,147237,214652,227319,103188,
           183866,222202,191383,191890,109873,
           133792,109860,156517,74715,69479,120718,
           68687,45030,37129,60886,62786,31732,28295,
           32148,40005,14809,11468,17749,17135,
           13005,6799,7717,9718,4810,3285,4249,
           3036,3287,1759,2402,1738,1010,2177,2063,
           1623,1730,1248,1895,2463,2276,3589,
           4195,2823,3450,4157,4570,2719,4083,6586,
           4617,5137,7796,6564,7405,7298,7867,
           7580,9771,11647,25827,25616,15632,10454,
           13278,16858,27550,18719,48277,28639,
           32971,20762,17972,18975,15609,18617
           )
)
```

Let's make a plot of the number of cases per year

```{r}
library(ggplot2)
```

```{r}
p <- ggplot(cdc) +
      aes(x=year,
          y=cases) +
      geom_line() +
      geom_point() +
      labs(x="Year",
           y="Cases per year",
           title="US Pertussis cases over the last century (CDC)") +
      scale_y_continuous(labels=scales::label_comma()) +
  theme_minimal()
```

The first Pertussis vaccine (whole-Pertussis, "wP") was introduced in 1946. An updated, antigen vaccine ("aP") was introduced in 1996, which had fewer side-effects than the wP vaccine. Let's add colored lines on the plots to indicate the introductions of these vaccines.

```{r}
p + geom_vline(xintercept = 1946, color="red") + # intro of wP vax
  geom_vline(xintercept = 1996, color="blue") + # intro of aP vax
  annotate("text", x=1947, y=190000, label="wP introduction", size=4, hjust=0, color="red") +
  annotate("text", x=1997, y=70000, label="aP introduction", size=4, hjust=0, color="blue")
```

Why is this vaccine-preventable disease on the upswing? To answer this questions, we need information about the immunological mechanisms for responses to the wP and aP vaccines. Luckily, the CMI-PB project collects exactly these data from participants longitudinally to answer this question.

# Exploring the CMI-PB data

We will use the `jsonlite` package to download CMI-PB data from their API.

```{r}
library(jsonlite)

subject <- read_json("https://www.cmi-pb.org/api/subject", simplifyVector = T)
head(subject)
```

> Q4. How many wP and aP subjects are there?

```{r}
table(subject$infancy_vac)
```

> Q5. How many males and females are in the dataset?

```{r}
table(subject$biological_sex)
```

> **Q6.** What is the breakdown of race and biological sex (e.g. number of Asian females, White males etc...)

```{r}
table(subject$biological_sex, subject$race)
```

Read in the `specimen` table

```{r}
specimen <- read_json("http://cmi-pb.org/api/specimen", simplifyVector = T)
head(specimen)
```

To know whether a given `specimen_id` comes from an aP or wP individual we need to link (a.k.a. "join" or merge) our `specimen` and `subject` data frames. The excellent **dplyr** package (that we have used previously) has a family of `join()` functions that can help us with this common task:

> **Q9.** Complete the code to join `specimen` and `subject` tables to make a new merged data frame containing all specimen records along with their associated subject details:

```{r}
#| message: false
library(dplyr)
```

```{r}
meta <- inner_join(specimen, subject)
dim(meta)
head(meta)
```

# Analyzing IgG titers

> **Q10.** Now using the same procedure join `meta` with `titer` data so we can further analyze this data in terms of time of visit aP/wP, male/female etc

```{r}
titer <- read_json("http://cmi-pb.org/api/ab_titer", simplifyVector = T)
head(titer)
```

```{r}
abdata <- inner_join(meta, titer)
dim(meta)
head(meta)
```

> **Q11.** How many specimens (i.e. entries in `abdata`) do we have for each `isotype`?

```{r}
table(abdata$isotype)
```

> **Q12.** What do you notice about the number of visit 8 specimens compared to other visits?

```{r}
table(abdata$visit, abdata$isotype)
```

There are many fewer antibody specimens for visit 8.

> Q. How many different antigens are there?

```{r}
length(unique(abdata$antigen))
table(abdata$antigen)
```

> **Q13.** Make a summary boxplot of Ab titer levels (MFI) for all antigens

```{r}
ig1 <- abdata %>% filter(isotype == "IgG1", visit!=8)
head(ig1)
```

```{r}
library(ggridges)
```

```{r}
ggplot(ig1) +
  aes(MFI, antigen) +
  geom_boxplot(alpha=0.4) + 
  facet_wrap(vars(visit), nrow=2) +
  theme(
    axis.text.x = element_text(angle=45, hjust=1)
  )
```
> Q14. What antigens show differences in the level of IgG1 antibody titers recognizing them over time? Why these and not others?

The biggest responders are FIM2/3 and FHA, for the most part. PRN may have some changes in reponse, but it could also very easily be classified as noise. 



```{r}
ggplot(ig1) +
  aes(MFI, antigen, col=infancy_vac) +
  geom_density_ridges2(alpha=0.4) + 
  facet_wrap(vars(visit), nrow=2) +
  theme(
    axis.text.x = element_text(angle=45, hjust=1)
  )
```
```{r}
ggplot(ig1) +
  aes(MFI, antigen, col=infancy_vac ) +
  geom_boxplot(show.legend = FALSE) + 
  facet_wrap(vars(infancy_vac, visit), nrow=2) +
  theme(
    axis.text.x = element_text(angle=45, hjust=1)
  )
```

Doesn't seem to be a dramatic difference in humoral response by vaccine type... there must be something else explaining our phenomenon!

> Q15. Filter to pull out only two specific antigens for analysis and create a boxplot for each. You can chose any you like. Below I picked a “control” antigen (“Measles”, that is not in our vaccines) and a clear antigen of interest (“FIM2/3”, extra-cellular fimbriae proteins from B. pertussis that participate in substrate attachment

```{r}
filter(ig1, antigen=="Measles") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
  geom_boxplot(show.legend = T) +
  facet_wrap(vars(visit)) +
  theme(
    axis.text.x = element_text(angle=45, hjust=1)
  ) +
  labs(title="Measles")

filter(ig1, antigen=="FIM2/3") %>%
  ggplot() +
  aes(MFI, col=infancy_vac) +
  geom_boxplot(show.legend = T) +
  facet_wrap(vars(visit)) +
  theme(
    axis.text.x = element_text(angle=45, hjust=1)
  ) +
  labs(title="FIM2/3")
```

# Obtaining CMI-PB RNASeq data
```{r}
url <- "https://www.cmi-pb.org/api/v2/rnaseq?versioned_ensembl_gene_id=eq.ENSG00000211896.7"
rna <- read_json(url, simplifyVector = T)
head(rna)
```

Let's join the RNA-seq data to our metadata
```{r}
ssrna <- inner_join(rna, meta)
head(ssrna)
```
> Q18. Make a plot of the time course of gene expression for IGHG1 gene (i.e. a plot of visit vs. tpm

```{r}
ggplot(ssrna) +
  aes(visit, tpm, group=subject_id) +
  geom_point() +
  geom_line(alpha=0.2)
```

> Q19. What do you notice about the expression of this gene?

Expression peaks for one visit and seems to fall back to normal.

> Q20. Does this pattern in time match the trend of antibody titer data? If not, why not?

It precedes the rise in IgG1 titer levels. This makes sense, considering you need antibody transcription before protein expression. Pretty remarkable how long the antibodies stick around though!

Let's look at this by vaccine status.
```{r}
ggplot(ssrna) +
  aes(tpm, col=infancy_vac) +
  geom_boxplot() +
  facet_wrap(vars(visit)) +
  labs(title="RNA expression by vaccine type and visit")
```

```{r}
ssrna %>%  
  filter(visit==4) %>% 
  ggplot() +
    aes(tpm, col=infancy_vac) + geom_density() + 
    geom_rug() +
  labs(title="RNA-expression at visit 4")
```























